### DISCUSSION TITLE
instant search?

### DISCUSSION THREAD

**Discussion by tlambert03** [chanzuckerberg/napari-hub#135] [Ideas]
Don't know anything about how the search is implemented here... but instant search (ala algolia) might be a nice feature for the search field.  Doesn't seem like we should have too much data for that sort of thing for quite a while?

**Comment by codemonkey800** [comment]
Right now, we currently use [fuse.js](https://fusejs.io/) for the search engine, and it's possible to perform instant search using it, but we intentionally made it so you had to submit the query before a search is executed. @liaprins-czi Any opinions on the UX of instant search and if it was something we should consider? 

Also, Algolia would be amazing to have. I've seen it used in a lot of popular JavaScript documentation sites (for example [Docusaurus](https://docusaurus.io/) and [Material UI](https://material-ui.com/)). @neuromusic @ziyangczi If/When we want to move to backend search, it may be worth looking into in the future (assuming we go with instant search). It looks like they have an [open source discount](https://www.algolia.com/for-open-source/) 🤔

**Comment by tlambert03** [comment]
yeah, I use algolia on one of my site (example search field [here](https://www.fpbase.org/)... type something like "mcherry")... I absolutely love it... incredibly performant.  great API.  I haven't paid a cent yet for my open source project...

**Comment by liaprins-czi** [comment]
> Any opinions on the UX of instant search and if it was something we should consider?

I like the idea of instant search in general (assuming by this term "instant search" we mean that the results begin to narrow as soon as user starts typing?). The main thing to consider is if there would be a lag at all, especially to include the highlighted snippets, which could be disorienting if so.

The other thing is that with this type of search they don't have to hit enter or click the search icon. This means the "x" button to clear results does not appear in place of the search icon. However, they can still just manually delete their input or we could change the interaction to be that the "x" appears next to the search icon instead of in place of it, as soon as there is at least one character typed.

**Comment by tlambert03** [comment]
> The main thing to consider is if there would be a lag at all

these things are pretty performant.  have a look through some of the examples here:
https://www.algolia.com/doc/guides/building-search-ui/resources/demos/react/

> This means the "x" button to clear results does not appear in place of the search icon. However, they can still just manually delete their input or we could change the interaction to be that the "x" appears next to the search icon instead of in place of it, as soon as there is at least one character typed.

yeah, this is all customizable

**Comment by neuromusic** [comment]
Looks like Algolia has a few other features that could be quite powerful...

- [search analytics](https://www.algolia.com/doc/guides/getting-insights-and-analytics/search-analytics/out-of-the-box-analytics/) (e.g. "top searches" and "searches that yield no results")
- filtering, sorting, faceting
- pagination

To bring on a new vendor like Algolia, we need to go through a rather rigorous audit of the vendor's legal terms and security practices (even at the free tier), so I don't think it would be worth it to pursue just for the instant search, since we can already do this with fuse, we'd need to migrate the indexing, etc.

However, as we look to adding features next half like filtering/navigating by categories of plugins, landing page improvements, pagination, etc, it might be worth it. If it meets design, engineering, and product requirements, and we want to use it, we can get it approved as a vendor.
